exercises/pig-latin/src/example.c

```diff
@@ -0,0 +1,121 @@
+#include "pig_latin.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdbool.h>
+
+static bool is_vowel(const char a)
+{
+   return a == 'a' || a == 'e' || a == 'i' || a == 'o' || a == 'u';
+}
+
+static bool starts_with_vowel(const char a, const char b)
+{
+   return is_vowel(a) || (a == 'x' && b == 'r') || (a == 'y' && b == 't');
+}
+
+static void count_words_and_chars(const char *input, size_t * words,
+                                  size_t * chars)
+{
+   bool in_word = false;
+   while (*input) {
+      (*chars)++;
+      if (isalpha(*input++)) {
+         *words += in_word ? 0 : 1;
+         in_word = true;
+      } else {
+         in_word = false;
+      }
+   }
+}
+
+static void mstrncpy(char *dst, const char *src, size_t n)
+{
+   while ((n--) && (*dst++ = tolower(*src++))) ;
+}
+
+static void append_ay(char *res)
+{
+   *res = 'a';
+   *(res + 1) = 'y';
+}
+
+static void process_pig_logic(char *dst, const char *src, size_t word_len)
+{
+   append_ay(dst + word_len);
+
+   const char a = tolower(*src);
+   const char b = a ? tolower(*(src + 1)) : 0;
+
+   if (starts_with_vowel(a, b)) {
+      mstrncpy(dst, src, word_len);
+      return;
+   }
+
+   size_t i = 0;
+   while (src[i]) {
+      const char a = tolower(src[i]);
+      const char b = tolower(src[i + 1]);
+
+      if (is_vowel(a))
+         break;
+
+      if (b == 'y') {
+         i++;
+         break;
+      }
+
+      if (a == 'q' && b == 'u') {
+         i += 2;
+         break;
+      }
+      i++;
+   }
+   mstrncpy(dst, src + i, word_len - i);
+   mstrncpy(dst + (word_len - i), src, i);
+}
+
+char *translate(const char *phrase)
```

Comment:
IMHO, this should not need to allocate memory - instead just alter the input, maybe make the input parameter `char phrase[]`. The returned length is the same as the input length so just changing the input makes memory handling a lot simpler.

Personally I'm quite averse to making functions that allocate memory but rely on the caller to `free` it. In real life, the function documentation would mention that it altered the input and the caller could do the allocation and make a copy if it needs the original to remain intact. This again looks to the single responsibility principle, if you can, make your class/function/etc take care of one thing as long as the functionality doesn't _need_ memory allocation then leave it to the caller to handle (in fact the caller might want to call some other code or library to do that for it).